### ISSUE TITLE
gc crash test case

### ISSUE THREAD

**Issue by erg** [slavapestov/factor#1401]
```factor
IN: scratchpad
: make-vocabs ( n -- vocabs )
    iota [ number>string "vocab" prepend create-vocab ] map
    [
        [ name>> set-current-vocab 1000 iota [ drop gensym [ + ] ( a b -- c ) define-declared ] each ] each
    ] keep ;
1000 [ [ 200 make-vocabs [ name>> forget-vocab ] each ] with-compilation-unit gc ] times
"scratchpad" set-current-vocab
```

```
Thread 0 Crashed:: Dispatch queue: com.apple.main-thread
0   libsystem_kernel.dylib        	0x00007fff8e467286 __pthread_kill + 10
1   libsystem_c.dylib             	0x00007fff90b67b53 abort + 129
2   org.factorcode.Factor         	0x0000000100002d14 factor::abort() + 84
3   org.factorcode.Factor         	0x00000001000143fd factor::factor_vm::allot_code_block(unsigned long, factor::code_block_type) + 317
4   org.factorcode.Factor         	0x0000000100014678 factor::factor_vm::add_code_block(factor::code_block_type, unsigned long, unsigned long, unsigned long, unsigned long, unsigned long, unsigned long, unsigned long) + 632 (code_blocks.cpp:420)
5   org.factorcode.Factor         	0x000000010002b516 factor::jit::to_code_block(unsigned long) + 166 (jit.cpp:128)
6   org.factorcode.Factor         	0x000000010003281c factor::factor_vm::jit_compile_quotation(unsigned long, unsigned long, bool) + 492 (quotations.cpp:314)
7   org.factorcode.Factor         	0x0000000100037139 factor::factor_vm::jit_compile_word(unsigned long, unsigned long, bool) + 233 (code_blocks.hpp:71)
8   org.factorcode.Factor         	0x0000000100037526 factor::factor_vm::allot_word(unsigned long, unsigned long, unsigned long) + 438 (tagged.hpp:32)
9   org.factorcode.Factor         	0x00000001000375d0 factor::factor_vm::primitive_word() + 48 (tagged.hpp:4)
10  ???                           	0x0000000126dc8511 0 + 4946953489
```

**Comment by erg** [comment]
Lame, it's just an out of memory error?

I'm calling forget though, shouldn't it never run out of memory?

```
Code heap used: 67103840
Code heap free: 928
fatal_error: Out of memory in add-compiled-block: 0

Program received signal SIGABRT, Aborted.
0x00007fff8e467286 in __pthread_kill ()
(gdb) bt
#0  0x00007fff8e467286 in __pthread_kill ()
#1  0x00007fff8a70642f in pthread_kill ()
#2  0x00007fff90b67b53 in abort ()
#3  0x0000000100003148 in factor::abort () at vm/os-unix.cpp:512
#4  0x00000001000396db in factor::fatal_error (msg=0x10006d392 "Out of memory in add-compiled-block", tagged=0) at vm/errors.cpp:22
#5  0x000000010002144a in factor::factor_vm::allot_code_block (this=0x100805200, size=48, type=factor::code_block_unoptimized) at vm/code_blocks.cpp:397
#6  0x000000010002159b in factor::factor_vm::add_code_block (this=0x100805200, type=factor::code_block_unoptimized, code_=4906827097, labels_=1, owner_=4906826700, relocation_=4906826841, parameters_=4906826898, literals_=4906826994, frame_size_untagged=32) at vm/code_blocks.cpp:420
#7  0x000000010004a83d in factor::jit::to_code_block (this=0x132d849e0, frame_size=32) at vm/jit.cpp:128
#8  0x0000000100056558 in factor::factor_vm::jit_compile_quotation (this=0x100805200, owner_=4906826700, quot_=4320402340, relocating=true) at vm/quotations.cpp:314
#9  0x000000010005efc3 in factor::factor_vm::jit_compile_word (this=0x100805200, word_=4906826700, def_=4320402340, relocating=true) at vm/words.cpp:18
#10 0x000000010005f5f3 in factor::factor_vm::allot_word (this=0x100805200, name_=4334775403, vocab_=1, hashcode_=5473136) at vm/words.cpp:60
#11 0x000000010005f6f2 in factor::factor_vm::primitive_word (this=0x100805200) at vm/words.cpp:71
#12 0x0000000100054b15 in primitive_word (parent=0x100805200) at vm/primitives.cpp:10
#13 0x0000000125d76061 in ?? ()
Previous frame inner to this frame (gdb could not unwind past this frame)
```

**Comment by mrjbq7** [comment]
Maybe forget is not cleaning everything up -- memory leak?

**Comment by bjourne** [comment]
I think when you forget stuff, then you forget methods which sits on generic methods. And when that happens the generic must be recompiled which causes a huge cascade because every word using that generic then also has to be recompiled. 

    Code heap used: 67103840
    Code heap free: 928
    fatal_error: Out of memory in add-compiled-block: 0

That's only 67mb. I think it should try and grow the code heap here.

**Comment by bjourne** [comment]
@erg is the crash reproducible? I'm running your make-vocabs loop for a long time, but I can't get it to ever run out of memory.

**Comment by erg** [comment]
It crashed for me after less than 5 minutes on Mac, but I'm not sure the
platform matters for this issue. Maybe it crashes sooner on 32bit?

**Comment by bjourne** [comment]
That must be it. I'm running 64bit but I'll see if it crashes faster on 32bit.

**Comment by bjourne** [comment]
Crashes after a long while:
```factor
IN: scratchpad 1000 [ [ 200 make-vocabs [ name>> forget-vocab ] each ] with-compilation-unit gc ] times
No suitable arithmetic method
left    f
right   0
generic -

Type :help for debugging help.
IN: vocab199 :c
...
(O) Word: (build-cfg)
(O) Word: begin-basic-block
(O) Word: end-local-analysis
(O) Word: local-kill-set
(O) Word: kill-locations
(O) Word: -
(O) Word: no-math-method
(O) Method: M\ object throw
(U) Quotation: [
        OBJ-CURRENT-THREAD special-object error-thread set-global
        current-continuation -> error-continuation set-global
        [ original-error set-global ] [ rethrow ] bi
    ]
```

**Comment by erg** [comment]
Fixed with the latest patch? Was it supposed to be fixed? (I'm trying it out)

**Comment by erg** [comment]
I let it run for like 10 minutes and it didn't crash. It used to crash pretty quickly. Maybe close this issue?

**Comment by bjourne** [comment]
It wasn't supposed to be fixed. But I've also run the loop for a long while and nothing crashes. So let's close this one since it's not reproducible anymore.

**Comment by erg** [comment]
The best issue for triggering a gc bug is still #1419 because it works first try in <1s on mac64 and depends on callstack size when it shouldnt. If you can get your hands on a mac, that is...

The best compiler bug to fix atm is #1370.

Closing because it doesn't crash anymore.
